evals/elsuite/tools/eval.py

```diff
@@ -0,0 +1,85 @@
+import json
+import logging
+from typing import Any, List
+from urllib.parse import parse_qs, urlparse
+
+from datasets import load_dataset
+from pydantic import BaseModel
+
+import evals
+import evals.metrics
+from evals.api import CompletionFn
+from evals.record import RecorderBase
+
+logger = logging.getLogger(__name__)
+DEFAULT_SAMPLE_RATE = 16000
+
+
+class Sample(BaseModel):
+    messages: List[dict[str, Any]]
+    functions: List[dict]
+
+
+class ToolsTask(evals.Eval):
+    def __init__(
+        self,
+        completion_fns: list[CompletionFn],
+        dataset: str,
+        temperature: float = 0.7,
+        max_tokens: int = 4096,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(completion_fns, *args, **kwargs)
+        assert len(completion_fns) == 1, "Audio tasks only support one completion fn"
+        self.dataset = dataset
+        self.temperature = temperature
+        self.max_tokens = max_tokens
+
+    def load_sample(self, row) -> list[Sample]:
+        messages = row["messages"]
+        last_tool_call_id = None
+        for m in messages:
+            if m["tool_calls"]:
+                last_tool_call_id = m["tool_calls"][-1]["id"]
+            elif m.get("tool_call_id") is not None and last_tool_call_id is not None:
+                m["tool_call_id"] = last_tool_call_id
+                last_tool_call_id = None
+
+        return Sample(messages=messages, functions=json.loads(row["functions"]))
+
+    def eval_sample(self, sample: Sample, rng):
+        assert isinstance(sample, Sample)
+        prompt = sample.messages[:2]
+        assistant_msg = sample.messages[2]
```

Comment:
This part likely needs work. One idea is to emit a sample on every assistant message.
A simpler version might be to emit only the first (or last) function call.

In some of the samples the model has to actually reject the tool call (e.g. if no tool for the job exists). I'll take a look to see if we can cover all cases, but I might go for a simpler first version.

Reply:
yeah I think this is right. Should be easy to test for a rejection, tool_calls=None in that case. But starting with only the first user message might be the right v1.

Reply:
The issue is there's not always a tool call after the first user message. Sometimes more info is needed. Though I guess then what we're measuring is whether the model called any function or not.

Reply:
Right. Maybe some of those cases are gray areas where the model just needed confirmation, in which case we'll need to be more liberal. But we can start simple and see how far we can get. 

I started looking into Gorilla, but it's really its own separate framework and will need to be augmented just like oaieval for audio data. For internal metrics, I think oaieval will probably be sufficient.

Reply:
Agreed.
I feel like these frameworks just do the same thing, just maybe with slightly different integrations.
Gorilla's metric is also just so simple. It's possible their data is useful, but I'm not sure about the whole framework.